In [124]:
# import pandas as pd
# df = pd.read_csv('IMDB Dataset.csv',on_bad_lines='warn')
# df.head()

import pandas as pd

# Attempt to read the file, handling potentially malformed lines
try:
    df = pd.read_csv('IMDB Dataset.csv', on_bad_lines='skip')  # Skip bad lines
except pd.errors.ParserError as e:
    print(f"Error encountered: {e}")
    # If skipping lines isn't enough, consider manual inspection
    # Print the problematic lines for further analysis
    with open('IMDB Dataset.csv', 'r') as f:
        for i, line in enumerate(f):
            if i in range(34065, 34070):  # Inspect lines around the error
                print(f"Line {i}: {line}")

# from google.colab import drive
# drive.mount('/content/drive')

# file_path = "/content/drive/My Drive/IMDB Dataset.csv"
# df = pd.read_csv(file_path)


##Converting to lower


In [125]:
import pandas as pd
df.head()

review sentiment
0  One of the other reviewers has mentioned that ...  positive
1  A wonderful little production. <br /><br />The...  positive
2  I thought this was a wonderful way to spend ti...  positive
3  Basically there's a family where a little boy ...  negative
4  Petter Mattei's "Love in the Time of Money" is...  positive

In [126]:
# df['review'] =df['review'][3].lower()
df['review']=df['review'].str.lower()
df

review sentiment
0      one of the other reviewers has mentioned that ...  positive
1      a wonderful little production. <br /><br />the...  positive
2      i thought this was a wonderful way to spend ti...  positive
3      basically there's a family where a little boy ...  negative
4      petter mattei's "love in the time of money" is...  positive
...                                                  ...       ...
49995  i thought this movie did a down right good job...  positive
49996  bad plot, bad dialogue, bad acting, idiotic di...  negative
49997  i am a catholic taught in parochial elementary...  negative
49998  i'm going to have to disagree with the previou...  negative
49999  no one expects the star trek movies to be high...  negative

[50000 rows x 2 columns]

## Remove stopwords

In [127]:
# df['review'].apply(remove_stopwords)

# Make sure the remove_stopwords function is defined before calling it
import nltk
from nltk.corpus import stopwords
nltk.download('stopwords')

def remove_stopwords(text4):
    new_text = []
    for word in text4.split():
        if word in stopwords.words('english'):
            new_text.append('')
        else:
            new_text.append(word)
    x = new_text[:]
    new_text.clear()
    return " ".join(x)

# Now you can

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [129]:
df.shape

(50000, 2)

##Remove HTML Tags

In [128]:
import re    #Regular expression (regex)
def removal_html_tags(text):
  pattern = re.compile('<.*?>')
  return pattern.sub(r'',text)

In [122]:
text ="<html><head><title>Sample HTML Page</title></head><body><h1>Welcome to My Website</h1><p>This is a <b>bold</b> paragraph with <i>italic</i> text.</p><h2>About Me</h2> <p>My name is John Doe, and I am a <a href=https://example.com>web developer</a>.</p><h3>My Skills</h3><ul><li>HTML & CSS</li><li>JavaScript</li> <li>Python</li><li>SQL</li></ul><h3>Contact Me</h3><p>Email: <a href=mailto:johndoe@example.com>johndoe@example.com</a></p><p>Phone: <span>+123 456 7890</span></p><footer><p>&copy; 2025 John Doe. All rights reserved.</p> </footer></body></html>"



In [130]:
removal_html_tags(text)

'Sample HTML PageWelcome to My WebsiteThis is a bold paragraph with italic text.About Me My name is John Doe, and I am a web developer.My SkillsHTML & CSSJavaScript PythonSQLContact MeEmail: johndoe@example.comPhone: +123 456 7890&copy; 2025 John Doe. All rights reserved. '

In [131]:
df['review'].apply(removal_html_tags)



0        one of the other reviewers has mentioned that ...
1        a wonderful little production. the filming tec...
2        i thought this was a wonderful way to spend ti...
3        basically there's a family where a little boy ...
4        petter mattei's "love in the time of money" is...
                               ...                        
49995    i thought this movie did a down right good job...
49996    bad plot, bad dialogue, bad acting, idiotic di...
49997    i am a catholic taught in parochial elementary...
49998    i'm going to have to disagree with the previou...
49999    no one expects the star trek movies to be high...
Name: review, Length: 50000, dtype: object

### Remove URLs



In [132]:
def remove_url(text2):
  pattern = re.compile(r'https?://\S+|www\.\S+')
  return pattern.sub(r'',text2)

##Remove Punctuation


In [133]:
 import string, time
 string.punctuation

'!"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~'

In [134]:
exclude = string.punctuation

In [135]:
def remove_punc(text2):
  for char in exclude:
    text2 = text2.replace(char, '')
  return text2


In [136]:
text2 = 'string. With. Punctuation?'

In [137]:
start = time.time()
print(remove_punc(text2))
time1 = time.time() - start
print(time1)

string With Punctuation
0.00013208389282226562


##Standard way to remove punctuation


In [138]:
def remove_punc(text2):
  return text2.translate(str.maketrans('', '', exclude))

In [139]:
start = time.time()
print(remove_punc(text2))
time2 = time.time() - start
print(time2)

string With Punctuation
0.00012421607971191406


In [140]:
time1/time2

1.0633397312859885

##Chat word treatment

In [141]:
chat_words = {'AFAIK':'As Far As I Know',
'AFK':'Away From Keyboard',
'ASAP':'As Soon As Possible',
'ATK':'At The Keyboard',
'ATM':'At The Moment',
'A3':'Anytime, Anywhere, Any'}

In [142]:
def chat_conversation(text3):
  new_text = []
  for w in text3.split():
    if w.upper() in chat_words:
      new_text.append(chat_words[w.upper()])
    else:
      new_text.append(w)
  return " ".join(new_text)

In [145]:
chat_conversation("ATM I am working")

'At The Moment I am working'

##Spelling Correction

In [147]:
!pip install textblob
from textblob import TextBlob
incorrect_text = "ceertain conditionas duriing seveal ggenerations aree moodified in the saame maner."
textBlb = TextBlob(incorrect_text)
textBlb.correct().string

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 624.3/624.3 kB 6.2 MB/s eta 0:00:00


'certain conditions during several generations are modified in the same manner.'

##Removing Stopwords

In [148]:
import nltk
from nltk.corpus import stopwords
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

In [149]:
stopwords.words('english')

['a',
 'about',
 'above',
 'after',
 'again',
 'against',
 'ain',
 'all',
 'am',
 'an',
 'and',
 'any',
 'are',
 'aren',
 "aren't",
 'as',
 'at',
 'be',
 'because',
 'been',
 'before',
 'being',
 'below',
 'between',
 'both',
 'but',
 'by',
 'can',
 'couldn',
 "couldn't",
 'd',
 'did',
 'didn',
 "didn't",
 'do',
 'does',
 'doesn',
 "doesn't",
 'doing',
 'don',
 "don't",
 'down',
 'during',
 'each',
 'few',
 'for',
 'from',
 'further',
 'had',
 'hadn',
 "hadn't",
 'has',
 'hasn',
 "hasn't",
 'have',
 'haven',
 "haven't",
 'having',
 'he',
 "he'd",
 "he'll",
 'her',
 'here',
 'hers',
 'herself',
 "he's",
 'him',
 'himself',
 'his',
 'how',
 'i',
 "i'd",
 'if',
 "i'll",
 "i'm",
 'in',
 'into',
 'is',
 'isn',
 "isn't",
 'it',
 "it'd",
 "it'll",
 "it's",
 'its',
 'itself',
 "i've",
 'just',
 'll',
 'm',
 'ma',
 'me',
 'mightn',
 "mightn't",
 'more',
 'most',
 'mustn',
 "mustn't",
 'my',
 'myself',
 'needn',
 "needn't",
 'no',
 'nor',
 'not',
 'now',
 'o',
 'of',
 'off',
 'on',
 'once',
 'on

In [150]:
def remove_stopwords(text4):
  new_text = []
  for word in text4.split():
    if word in stopwords.words('english'):
      new_text.append('')
    else:
      new_text.append(word)
  x = new_text[:]
  new_text.clear()
  return " ".join(x)

In [151]:
remove_stopwords('probably my all time favourite movie, a story of selfishness, sacrifice and dedication toa noble cause')

'probably   time favourite movie,  story  selfishness, sacrifice  dedication toa noble cause'

##Handling emojis


In [152]:
import re
def remove_emoji(text5):
  emoji_pattern = re.compile("["
                          u"\U0001F600-\U0001F64F"  # emoticons
                          u"\U0001F300-\U0001F5FF"  # symbols & pictographs
                          u"\U0001F680-\U0001F6FF"  # transport & map symbols
                          u"\U0001F1E0-\U0001F1FF"  # flags (iOS)
                          u"\U00002702-\U000027B0"
                          u"\U000024C2-\U0001F251"
                          "]+", flags=re.UNICODE)
  return emoji_pattern.sub(r'', text5)

In [153]:
remove_emoji("game is on 🔥🔥")

'game is on '

In [154]:
!pip install emoji
import emoji
print(emoji.demojize('game is on 🔥'))

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 590.6/590.6 kB 6.8 MB/s eta 0:00:00
game is on :fire:


##Tokenization


**- Using the split function**


In [155]:
#word tokenization
sent1 = 'I am going to delhi'
sent1.split()

['I', 'am', 'going', 'to', 'delhi']

In [156]:
#sentence tokenization
sent2 = 'I am going to delhi. I will stay there for 3 days. Let\'s hope the trip to be great'
sent2.split('.')


['I am going to delhi',
 ' I will stay there for 3 days',
 " Let's hope the trip to be great"]

In [157]:
#Problem with split function
sent3 = 'I am going to delhi!'
sent3.split()

['I', 'am', 'going', 'to', 'delhi!']

In [158]:
sent4= 'Where do think I should go?I have 3 day holiday'
sent4.split('.')

['Where do think I should go?I have 3 day holiday']

**- Regular Expression**

In [159]:
import re
sent3 = 'I am going to delhi!'
tokens = re.findall("[\w']+", sent3)
tokens

['I', 'am', 'going', 'to', 'delhi']

In [160]:
text = """Lorem Ipsum is simply dummy text of the printing and typesetting industry?
Lorem Ipsum has been the industry's standard dummy text ever since the 1500s,
when an unknown printer took a galley of type and scrambled it to make a type specimen book.
It has survived not only five centuries, but also the leap into."""
sentences = re.compile('[.!?] ').split(text)
sentences

["Lorem Ipsum is simply dummy text of the printing and typesetting industry?\nLorem Ipsum has been the industry's standard dummy text ever since the 1500s,\nwhen an unknown printer took a galley of type and scrambled it to make a type specimen book.\nIt has survived not only five centuries, but also the leap into."]

**- NLTK**

In [161]:
import nltk
nltk.download('punkt_tab')
from nltk.tokenize import word_tokenize,sent_tokenize
sent1 = 'I am going to visit delhi!'
word_tokenize(sent1)

[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!


['I', 'am', 'going', 'to', 'visit', 'delhi', '!']

In [162]:
sent5 = 'I have a Ph.D in A.I'
sent6 = "We're here to help! mail us john.c.breckinridge@altostrat.com"
sent7 = 'A 5km ride cost $10.50'
word_tokenize(sent5)


['I', 'have', 'a', 'Ph.D', 'in', 'A.I']

In [163]:
word_tokenize(sent6)

['We',
 "'re",
 'here',
 'to',
 'help',
 '!',
 'mail',
 'us',
 'john.c.breckinridge',
 '@',
 'altostrat.com']

In [164]:
word_tokenize(sent7)

['A', '5km', 'ride', 'cost', '$', '10.50']

**- Spacy**

In [165]:
import spacy
nlp = spacy.load('en_core_web_sm')

In [166]:
doc1 = nlp(sent5)
doc2 = nlp(sent6)
doc3 = nlp(sent7)
doc4 = nlp(sent1)

In [167]:
for token in doc2:
  print(token)

We
're
here
to
help
!
mail
us
john.c.breckinridge@altostrat.com


##Stemming


In [168]:
#Stemming is used in Information retrieval system
#do -> un do able
#walk -> walking,walked, walks


In [169]:
from nltk.stem.porter import PorterStemmer

In [170]:
ps = PorterStemmer()
def stem_words(text7):
  return " ".join([ps.stem(word) for word in text7.split()])

sample = "walk walks walking walked"
stem_words(sample)

'walk walk walk walk'

In [171]:
text7 = 'I am meeting him tomorrow at the meeting'
stem_words(text7)

'i am meet him tomorrow at the meet'

##Lemmatization
- Slow compared to Stemming but gives the correct english output word
- wordnet Lemmatization - Lexical dictionary

In [172]:
import nltk
from nltk.stem import WordNetLemmatizer
nltk.download('wordnet')

wordnet_lemmatizer = WordNetLemmatizer()

sentence = "He was running and eating at same time. He has bad habit of swimming after playing long hours in the Sun."
punctuations = "?:!.,;"
sentence_words = nltk.word_tokenize(sentence)
for word in sentence_words:
  if word in punctuations:
    sentence_words.remove(word)

sentence_words
print("{0:20}{1:20}".format("Word","Lemma"))
for word in sentence_words:
  print("{0:20}{1:20}".format(word,wordnet_lemmatizer.lemmatize(word,pos='v')))


[nltk_data] Downloading package wordnet to /root/nltk_data...


Word                Lemma               
He                  He                  
was                 be                  
running             run                 
and                 and                 
eating              eat                 
at                  at                  
same                same                
time                time                
He                  He                  
has                 have                
bad                 bad                 
habit               habit               
of                  of                  
swimming            swim                
after               after               
playing             play                
long                long                
hours               hours               
in                  in                  
the                 the                 
Sun                 Sun                 


##Plan of Attack

* Bag of words
* TF/IDF
* One Hot Encoding
* ngrams
* Custom feature
* Word2Vec -> Embedding

_________________________________________

- What is the Feature Extraction from text?
- Why do we need it?
- Why is it difficult?
- What is the core idea?
- What are the techniques?


#One - Hot encoding is not used why?
- Sparsity
- No fixed size
- OUT OF VOCABULARY (OOV)
- No capturing of Semantic meaning
- Semantic meaning is lost


##Bag of words
 - Used more in text classification
Disadvantages:
- Sparsity
- Out of Vocab(OOV)
- Ordering


In [173]:
import numpy as np
import pandas as pd


In [174]:
df = pd.DataFrame({
    'text':['people watch campusx','campusx watch campusx','people write comment','campusx write comment'],'output':[1,1,0,0]})

In [175]:
df

text  output
0   people watch campusx       1
1  campusx watch campusx       1
2   people write comment       0
3  campusx write comment       0

In [176]:
from sklearn.feature_extraction.text import CountVectorizer
cv = CountVectorizer()

In [177]:
#Bag of words
bow = cv.fit_transform(df['text'])

In [178]:
#Vocab( named in t form of index)
print(cv.vocabulary_)

{'people': 2, 'watch': 3, 'campusx': 0, 'write': 4, 'comment': 1}


In [179]:
#to convert into the numpy array
print(bow[0].toarray())  #first sentence
print(bow[1].toarray())  #second sentence

[[1 0 1 1 0]]
[[2 0 0 1 0]]


In [180]:
#when new sentence arrives
cv.transform(['campusx watch and write comment of campusx']).toarray()

array([[2, 1, 0, 1, 1]])

##  Bag of N-grams

- #Advantages:
- Able to capture Semantic meaning of the sentence
- Easy to implement

-#Disadvantages:
- Dimension of the vocabulary increases
- Out of Vocabulary



In [181]:
from sklearn.feature_extraction.text import CountVectorizer
# cv = CountVectorizer(ngram_range=(2,2))
cv = CountVectorizer(ngram_range=(1,2)) # contains both uni and bi grams

In [182]:
bow = cv.fit_transform(df['text'])
print(cv.vocabulary_)

{'people': 4, 'watch': 7, 'campusx': 0, 'people watch': 5, 'watch campusx': 8, 'campusx watch': 1, 'write': 9, 'comment': 3, 'people write': 6, 'write comment': 10, 'campusx write': 2}


#TF-IDF

In [183]:
df


text  output
0   people watch campusx       1
1  campusx watch campusx       1
2   people write comment       0
3  campusx write comment       0

In [184]:
from sklearn.feature_extraction.text import TfidfVectorizer
tfidf = TfidfVectorizer()
tfidf.fit_transform(df['text']).toarray()

array([[0.49681612, 0.        , 0.61366674, 0.61366674, 0.        ],
       [0.8508161 , 0.        , 0.        , 0.52546357, 0.        ],
       [0.        , 0.57735027, 0.57735027, 0.        , 0.57735027],
       [0.49681612, 0.61366674, 0.        , 0.        , 0.61366674]])

In [185]:
print(tfidf.idf_)
print(tfidf.get_feature_names_out())

[1.22314355 1.51082562 1.51082562 1.51082562 1.51082562]
['campusx' 'comment' 'people' 'watch' 'write']


## Word2Vec
 Advantages:
 - Captures the Semantic meaning of words
 - Low dimension vector -> computation speed increases
 - Dense vector -> non zeros -reduces overfitting
 ----------------------------------------
 - Increase the training
 - Increase the dimension of vectors
 - Increase window size



Word2Vec -> two words sharing similar contexts also share a similar meaning and consequently a similar vector representation from the model.

- We will use pre-trained weights of word2vec that was trained on Google New corpus containing 3 billion word. The model consists of 300-dimensional vectors for 3 billion words and phrases.

In [186]:
  import gensim
  from gensim.models import Word2Vec,KeyedVectors

In [187]:
!pip install wget

In [ ]:
# !wget -c "https://s3.amazonaws.com/dl4j-distribution/GoogleNews-vectors-negative300.bin.gz"
!wget -c "https://s3.amazonaws.com/dl4j-distribution/GoogleNews-vectors-negative300.bin.gz" # Re-download with resume option
!ls -lh GoogleNews-vectors-negative300.bin.gz # Check the file size

--2025-02-20 10:14:13--  https://s3.amazonaws.com/dl4j-distribution/GoogleNews-vectors-negative300.bin.gz
Resolving s3.amazonaws.com (s3.amazonaws.com)... 16.15.176.20, 16.182.103.136, 52.216.48.112, ...
Connecting to s3.amazonaws.com (s3.amazonaws.com)|16.15.176.20|:443... connected.
HTTP request sent, awaiting response... 404 Not Found
2025-02-20 10:14:13 ERROR 404: Not Found.

-rw-r--r-- 1 root root 88M Feb 20 10:14 GoogleNews-vectors-negative300.bin.gz


In [ ]:
import gensim
from gensim.models import Word2Vec, KeyedVectors

model = KeyedVectors.load_word2vec_format('GoogleNews-vectors-negative300.bin', binary=True, limit=500000)

In [ ]:
model['cricket']

array([-3.67187500e-01, -1.21582031e-01,  2.85156250e-01,  8.15429688e-02,
        3.19824219e-02, -3.19824219e-02,  1.34765625e-01, -2.73437500e-01,
        9.46044922e-03, -1.07421875e-01,  2.48046875e-01, -6.05468750e-01,
        5.02929688e-02,  2.98828125e-01,  9.57031250e-02,  1.39648438e-01,
       -5.41992188e-02,  2.91015625e-01,  2.85156250e-01,  1.51367188e-01,
       -2.89062500e-01, -3.46679688e-02,  1.81884766e-02, -3.92578125e-01,
        2.46093750e-01,  2.51953125e-01, -9.86328125e-02,  3.22265625e-01,
        4.49218750e-01, -1.36718750e-01, -2.34375000e-01,  4.12597656e-02,
       -2.15820312e-01,  1.69921875e-01,  2.56347656e-02,  1.50146484e-02,
       -3.75976562e-02,  6.95800781e-03,  4.00390625e-01,  2.09960938e-01,
        1.17675781e-01, -4.19921875e-02,  2.34375000e-01,  2.03125000e-01,
       -1.86523438e-01, -2.46093750e-01,  3.12500000e-01, -2.59765625e-01,
       -1.06933594e-01,  1.04003906e-01, -1.79687500e-01,  5.71289062e-02,
       -7.41577148e-03, -

In [ ]:
model.most_similar('cricket')

[('cricketing', 0.8372225761413574),
 ('cricketers', 0.8165745735168457),
 ('Test_cricket', 0.8094819188117981),
 ('Twenty##_cricket', 0.8068488240242004),
 ('Twenty##', 0.7624265551567078),
 ('Cricket', 0.75413978099823),
 ('cricketer', 0.7372578382492065),
 ('twenty##', 0.7316356897354126),
 ('T##_cricket', 0.7304614186286926),
 ('West_Indies_cricket', 0.6987985968589783)]

In [ ]:
model.most_similar('facebook')

[('Facebook', 0.7563533186912537),
 ('FaceBook', 0.7076998949050903),
 ('twitter', 0.6988552212715149),
 ('myspace', 0.6941817998886108),
 ('Twitter', 0.664244532585144),
 ('Facebook.com', 0.6529868245124817),
 ('FacebookFacebook', 0.6162722110748291),
 ('facebook.com', 0.6135972142219543),
 ('Twitter.com', 0.6102108359336853),
 ('TwitterTwitter', 0.6085205674171448)]

In [ ]:
model.similarity('man','woman')

0.76640123

In [ ]:
model.doesnt_match(['PHP','java','monkey'])

'monkey'

In [ ]:
vec = model['king'] - model['man'] + model['woman']
model.most_similar([vec])

[('king', 0.8449392318725586),
 ('queen', 0.7300517559051514),
 ('monarch', 0.645466148853302),
 ('princess', 0.6156251430511475),
 ('crown_prince', 0.5818676352500916),
 ('prince', 0.5777117609977722),
 ('kings', 0.5613663792610168),
 ('sultan', 0.5376775860786438),
 ('queens', 0.5289887189865112),
 ('ruler', 0.5247419476509094)]

In [ ]:
vec = model['INR'] - model ['India'] + model['England']
model.most_similar([vec])

[('INR', 0.6442341208457947),
 ('GBP', 0.5040826797485352),
 ('England', 0.44649264216423035),
 ('£', 0.43340998888015747),
 ('Â_£', 0.4307197630405426),
 ('£_#.##m', 0.42561301589012146),
 ('Pounds_Sterling', 0.42512619495391846),
 ('GBP##', 0.42464491724967957),
 ('stg', 0.42324796319007874),
 ('£_#.###m', 0.4201711118221283)]

## Game-of-thrones-word2vec


In [ ]:
import numpy as np
import pandas as pd

In [38]:
!pip install gensim

In [ ]:
import gensim
import os
from nltk import sent_tokenize

In [41]:
# import nltk
# from nltk import sent_tokenize
# nltk.download('punkt_tab')
# from gensim.utils import simple_preprocess
# import os

# story =[]
# for filename in os.listdir('/content/sample_data/Data'):

#   f = open(os.path.join('/content/sample_data/Data',filename))
#   corpus = f.read()
#   raw_sent = sent_tokenize(corpus)
#   for sent in raw_sent:
#     story.append(simple_preprocess(sent))

import nltk
from nltk import sent_tokenize
nltk.download('punkt_tab')
from gensim.utils import simple_preprocess
import os

story = []
for filename in os.listdir('/content/sample_data/Data'):
    # Specify the encoding when opening the file
    with open(os.path.join('/content/sample_data/Data', filename), 'r', encoding='latin-1') as f:
        corpus = f.read()
        raw_sent = sent_tokenize(corpus)
        for sent in raw_sent:
            story.append(simple_preprocess(sent))



[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!


In [42]:
story

[['storm',
  'of',
  'swords',
  'book',
  'three',
  'of',
  'song',
  'of',
  'ice',
  'and',
  'fire',
  'by',
  'george'],
 ['martin',
  'note',
  'on',
  'chronology',
  'song',
  'of',
  'ice',
  'and',
  'fire',
  'is',
  'told',
  'through',
  'the',
  'eyes',
  'of',
  'characters',
  'who',
  'are',
  'sometimes',
  'hundreds',
  'or',
  'even',
  'thousands',
  'of',
  'miles',
  'apart',
  'from',
  'one',
  'another'],
 ['some',
  'chapters',
  'cover',
  'day',
  'some',
  'only',
  'an',
  'hour',
  'others',
  'might',
  'span',
  'fortnight',
  'month',
  'half',
  'year'],
 ['with',
  'such',
  'structure',
  'the',
  'narrative',
  'cannot',
  'be',
  'strictly',
  'sequential',
  'sometimes',
  'important',
  'things',
  'are',
  'happening',
  'simultaneously',
  'thousand',
  'leagues',
  'apart'],
 ['in',
  'the',
  'case',
  'of',
  'the',
  'volume',
  'now',
  'in',
  'hand',
  'the',
  'reader',
  'should',
  'realize',
  'that',
  'the',
  'opening',
  'chap

- Window Size (window=10): Determines how many words before and after a target word are considered for learning relationships.
- Minimum Count (min_count=2): Ignores words that appear less than twice in the dataset.

In [46]:
model = gensim.models.Word2Vec(
    window=10,
    min_count=2,

)

In [47]:
#to extract the unique words
model.build_vocab(story)

- Now, the model learns word relationships by training on story. It tries to adjust word vectors so that words appearing in similar contexts have similar embeddings.


**How Word2Vec Learns Word Relationships ?**
Word2Vec uses neural networks to learn word relationships based on context. It uses one of two approaches:

- CBOW (Continuous Bag of Words) → Predicts a word given its surrounding words.
- Skip-gram → Predicts surrounding words given a target word.
- Your model is using one of these techniques (by default, it's Skip-gram if not specified).



In [48]:
model.train(story,total_examples=model.corpus_count,epochs=model.epochs)

(6568882, 8628190)

In [49]:
model.wv.most_similar('daenerys')

[('stormborn', 0.8264977335929871),
 ('unburnt', 0.7213401198387146),
 ('queen', 0.6968168020248413),
 ('targaryen', 0.6944656372070312),
 ('princess', 0.6765398383140564),
 ('elia', 0.672277569770813),
 ('dorne', 0.6662358641624451),
 ('viserys', 0.6618695855140686),
 ('dragons', 0.6536297798156738),
 ('myrcella', 0.6478593945503235)]

In [66]:
model.wv.doesnt_match(['jon','rikon','robb','arya','sansa','bran'])

'jon'

In [54]:
model.wv['jon']

array([-0.30805373, -0.49379575, -0.87950355,  0.4494075 ,  1.1735126 ,
       -1.6462055 ,  1.3804729 ,  1.5015495 , -0.17569646,  0.85835034,
       -1.4574494 ,  0.20809379, -1.2477425 , -1.4664822 , -1.0972269 ,
        0.32044598,  0.96055305,  2.439985  , -1.5077548 , -2.510345  ,
        0.2174339 , -0.4302559 , -0.32082173,  1.9009798 , -0.49587414,
       -0.40456545, -0.44340488,  0.81090426, -1.3907816 ,  0.77607304,
       -2.3558803 , -0.6685452 ,  1.0229647 , -1.2347485 ,  0.3512909 ,
        1.1565443 ,  1.0596806 ,  0.3117768 , -0.29399207,  0.33361518,
        0.29400447, -0.71437967, -0.28628007,  2.097539  ,  1.6650176 ,
       -0.6174752 , -0.98045915, -0.20283169, -0.18439315,  0.96828896,
        0.66546535, -0.18824907,  1.28671   , -0.12198772,  0.26551425,
        0.38497463, -0.9524275 , -2.6155987 ,  0.32310772, -0.49682686,
        0.4316979 ,  0.04286176,  2.7163    , -1.6842829 ,  0.40347692,
       -0.71816766,  1.4451962 ,  2.3040116 , -1.0951225 , -1.57

In [56]:
model.wv.similarity('daenerys','jon')

-0.18165828

In [57]:
model.wv.similarity('arya','sansa')

0.8561988

In [58]:
model.wv.get_normed_vectors()

array([[ 0.03132202,  0.02544114,  0.0621711 , ..., -0.13539006,
        -0.0338271 ,  0.12474199],
       [-0.20909056, -0.06988856,  0.04717761, ..., -0.01472174,
        -0.08043273,  0.2152738 ],
       [ 0.21625002,  0.03754369, -0.18732867, ...,  0.03794577,
         0.10364863, -0.24082512],
       ...,
       [-0.11645076,  0.02091977,  0.04238984, ..., -0.11086711,
        -0.01276368, -0.0718834 ],
       [ 0.1501305 , -0.02131767,  0.02849175, ..., -0.04239932,
         0.01375146, -0.06325132],
       [-0.1713236 ,  0.05231297, -0.018894  , ...,  0.03726123,
        -0.00529968, -0.00185603]], dtype=float32)

In [188]:
model.wv.get_normed_vectors().shape
# 17,453 → The number of unique words (vocabulary size).
# 100 → The number of dimensions for each word vector.

# "King" might be represented as:
# [-0.12, 0.45, ..., 0.78]  # (100 numbers)

(31845, 100)

In [60]:
y= model.wv.index_to_key

In [61]:
y

['the',
 'and',
 'to',
 'of',
 'he',
 'his',
 'was',
 'you',
 'her',
 'in',
 'it',
 'had',
 'that',
 'she',
 'as',
 'with',
 'him',
 'not',
 'but',
 'for',
 'they',
 'is',
 'at',
 'on',
 'said',
 'my',
 'have',
 'be',
 'lord',
 'them',
 'no',
 'from',
 'would',
 'were',
 'me',
 'your',
 'one',
 'all',
 'when',
 'will',
 'ser',
 'if',
 'so',
 'their',
 'we',
 'could',
 'are',
 'man',
 'there',
 'this',
 'up',
 'been',
 'what',
 'did',
 'by',
 'king',
 'do',
 'men',
 'back',
 'out',
 'more',
 'or',
 'who',
 'down',
 'well',
 'than',
 'only',
 'like',
 'jon',
 'some',
 'father',
 'old',
 'hand',
 'even',
 'too',
 'tyrion',
 'before',
 'never',
 'an',
 'off',
 'know',
 'see',
 'into',
 'made',
 'now',
 'eyes',
 'black',
 'told',
 'lady',
 'thought',
 'time',
 'then',
 'how',
 'long',
 'has',
 'can',
 'might',
 'us',
 'come',
 'where',
 'here',
 'through',
 'face',
 'still',
 'head',
 'red',
 'll',
 'way',
 'boy',
 'page',
 'must',
 'once',
 'queen',
 'good',
 'two',
 'brother',
 'night',
 

**PCA for Dimensionality Reduction:**
Word embeddings are high-dimensional (usually 100–300 dimensions). PCA (Principal Component Analysis) reduces these to a smaller number of dimensions for visualization.

In [189]:
from sklearn.decomposition import PCA
pca = PCA(n_components=3)
X = pca.fit_transform(model.wv.get_normed_vectors())
X.shape

(31845, 3)

In [190]:
import plotly.express as px
fig = px.scatter_3d(X[200:300],x=0,y=1,z=2,color=y[200:300])
fig.show()

## IMDB Dataset -> Modeling

In [191]:
import pandas as pd
import numpy as np

# Use the 'error_bad_lines=False' argument to skip problematic rows
temp_df = pd.read_csv('IMDB Dataset.csv', on_bad_lines='skip')
df = temp_df.iloc[:10000]
df.head()

review sentiment
0  One of the other reviewers has mentioned that ...  positive
1  A wonderful little production. <br /><br />The...  positive
2  I thought this was a wonderful way to spend ti...  positive
3  Basically there's a family where a little boy ...  negative
4  Petter Mattei's "Love in the Time of Money" is...  positive

In [72]:
df['review'][1]

'A wonderful little production. <br /><br />The filming technique is very unassuming- very old-time-BBC fashion and gives a comforting, and sometimes discomforting, sense of realism to the entire piece. <br /><br />The actors are extremely well chosen- Michael Sheen not only "has got all the polari" but he has all the voices down pat too! You can truly see the seamless editing guided by the references to Williams\' diary entries, not only is it well worth the watching but it is a terrificly written and performed piece. A masterful production about one of the great master\'s of comedy and his life. <br /><br />The realism really comes home with the little things: the fantasy of the guard which, rather than use the traditional \'dream\' techniques remains solid then disappears. It plays on our knowledge and our senses, particularly with the scenes concerning Orton and Halliwell and the sets (particularly of their flat with Halliwell\'s murals decorating every surface) are terribly well d

In [73]:
df['sentiment'].value_counts()

sentiment
positive    5028
negative    4972
Name: count, dtype: int64

In [74]:
df.isnull().sum()

review       0
sentiment    0
dtype: int64

In [75]:
df.duplicated().sum()

17

In [76]:
df.drop_duplicates(inplace=True)

<ipython-input-76-16cdb8520be8>:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [77]:
df.duplicated().sum()

0

- Basic preprocessing
- Remove tags
- Lowercase
- Remove stopwords



In [78]:
import re
def remove_tags(raw_text):
  cleaned_text = re.sub(re.compile('<.*?>'),'',raw_text)
  return cleaned_text

In [79]:
df['review'] = df['review'].apply(remove_tags)

<ipython-input-79-00e93c2a3043>:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [80]:
df

review sentiment
0     One of the other reviewers has mentioned that ...  positive
1     A wonderful little production. The filming tec...  positive
2     I thought this was a wonderful way to spend ti...  positive
3     Basically there's a family where a little boy ...  negative
4     Petter Mattei's "Love in the Time of Money" is...  positive
...                                                 ...       ...
9995  Fun, entertaining movie about WWII German spy ...  positive
9996  Give me a break. How can anyone say that this ...  negative
9997  This movie is a bad movie. But after watching ...  negative
9998  This is a movie that was probably made to ente...  negative
9999  Smashing film about film-making. Shows the int...  positive

[9983 rows x 2 columns]

In [81]:
df['review'] = df['review'].apply(lambda x:x.lower())

<ipython-input-81-afca29351ee0>:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [82]:
from nltk.corpus import stopwords
nltk.download('stopwords')
sw_list = stopwords.words('english')
df['review'] = df['review'].apply(lambda x: [item for item in x.split() if item not in sw_list]).apply(lambda x:" ".join(x))

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.
<ipython-input-82-2f4050987a35>:4: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [83]:
df

review sentiment
0     one reviewers mentioned watching 1 oz episode ...  positive
1     wonderful little production. filming technique...  positive
2     thought wonderful way spend time hot summer we...  positive
3     basically there's family little boy (jake) thi...  negative
4     petter mattei's "love time money" visually stu...  positive
...                                                 ...       ...
9995  fun, entertaining movie wwii german spy (julie...  positive
9996  give break. anyone say "good hockey movie"? kn...  negative
9997  movie bad movie. watching endless series bad h...  negative
9998  movie probably made entertain middle school, e...  negative
9999  smashing film film-making. shows intense stran...  positive

[9983 rows x 2 columns]

##Purpose in ML:
- X (features): Independent variables (e.g., text data in NLP).
- y (target): Dependent variable (e.g., sentiment labels like "positive" or "negative").

In [84]:
X = df.iloc[:,0:1]
Y = df['sentiment']

In [85]:
X

review
0     one reviewers mentioned watching 1 oz episode ...
1     wonderful little production. filming technique...
2     thought wonderful way spend time hot summer we...
3     basically there's family little boy (jake) thi...
4     petter mattei's "love time money" visually stu...
...                                                 ...
9995  fun, entertaining movie wwii german spy (julie...
9996  give break. anyone say "good hockey movie"? kn...
9997  movie bad movie. watching endless series bad h...
9998  movie probably made entertain middle school, e...
9999  smashing film film-making. shows intense stran...

[9983 rows x 1 columns]

In [86]:
Y

0       positive
1       positive
2       positive
3       negative
4       positive
          ...   
9995    positive
9996    negative
9997    negative
9998    negative
9999    positive
Name: sentiment, Length: 9983, dtype: object

- The target variable (Y_train, Y_test) represents categories (e.g., Positive/Negative sentiment).

- If the labels are categorical (e.g., "Positive", "Negative"), we use Label Encoding to convert them to numbers.

**This does not require BoW, because labels are not text-based features.**



In [87]:
from sklearn.preprocessing import LabelEncoder
encoder = LabelEncoder()
Y = encoder.fit_transform(Y)

# LabelEncoder is used to convert
# categorical labels (text) into numerical values.

In [88]:
Y

array([1, 1, 1, ..., 0, 0, 1])

- This imports the train_test_split function from scikit-learn, which is used to split a dataset into training and testing sets.
- train_test_split -> Splits dataset into training and testing sets
- test_size=0.2	-> 20% test data, 80% training data
- random_state=1 -> Ensures consistent splitting for reproducibility
- X_train, Y_train	Used to train the model
- X_test, Y_test	Used to evaluate model performance

In [89]:
from sklearn.model_selection import train_test_split
X_train,X_test,Y_train,Y_test = train_test_split(X,Y,test_size=0.2,random_state=1)

In [90]:
X_train.shape

(7986, 1)

## Applying Bag of Words

In [91]:
from sklearn.feature_extraction.text import CountVectorizer
cv = CountVectorizer()

- X_train_bow = Bag of Words (BoW) representation of training reviews.
- X_test_bow = BoW representation of testing reviews.

### Why to take xtrain and xtest and not ytrain and ytest?
- Ans:
- Machine learning models do not understand text; they work with numerical representations.
- BoW converts text data into a numerical format (word counts or presence/absence of words).
- Only the features (X) need transformation, because they are the actual input to the model.

In [92]:
X_train_bow = cv.fit_transform(X_train['review']).toarray()
X_test_bow = cv.transform(X_test['review']).toarray()


- "Amazing movie, loved it!"	[0,1,1,0,0,1,...,0]
- "Worst film ever"	[1,0,0,0,1,0,...,0]
- "Fantastic and thrilling"	[0,1,0,1,0,0,...,0]


ANS: If there were 3 reviews and 6 unique words, the shape would be:
(3,6)
- In your case:
7986 reviews, 48282 unique words → (7986, 48282)

In [93]:
X_train_bow.shape

(7986, 48282)

## Training the model

- X_train_bow is the Bag of Words (BoW) representation of X_train.It is a numerical matrix where each row represents a review and each column represents a word count.
X_train_bow =
[
    [1, 0, 1, 0, 2],  # First review
    [0, 3, 1, 1, 0],  # Second review
    [2, 0, 0, 1, 1]   # Third review
]

- Y_train contains the actual sentiment labels (e.g., Positive or Negative).
Example:
Y_train = ["Positive", "Negative", "Positive"]
Y_train = [1, 0, 1]

**- GaussianNB() learns how X_train_bow maps to Y_train, enabling it to make future predictions.**

In [94]:
from sklearn.naive_bayes import GaussianNB
gnb = GaussianNB()
gnb.fit(X_train_bow,Y_train)

GaussianNB()

In [95]:
# y_pred = gnb.predict(X_test_bow)
# Uses the trained Naïve Bayes classifier (gnb) to predict the class labels for X_test_bow.
# X_test_bow is the Bag of Words representation of the test data.
# The result, y_pred, is an array containing
# predicted class labels (e.g., positive/negative or spam/not spam).

y_pred = gnb.predict(X_test_bow)
from sklearn.metrics import accuracy_score, confusion_matrix
accuracy_score(Y_test,y_pred)

0.6324486730095142

In [96]:
# [[TN  FP]
#  [FN  TP]]


confusion_matrix(Y_test,y_pred)

array([[717, 235],
       [499, 546]])

## To increase the accuracy -> Use RandomForestClassifier


-What is RandomForestClassifier()?
**Random Forest is an ensemble learning method that combines multiple Decision Trees to improve prediction accuracy and reduce overfitting.**

- Unlike Naïve Bayes (which assumes feature independence), Random Forest:
- Creates multiple Decision Trees on different subsets of data.
- Combines their outputs (majority vote for classification).
- Handles non-linearity and interactions between words better than Naïve Bayes.


In [97]:
from sklearn.ensemble import RandomForestClassifier
rf = RandomForestClassifier()
rf.fit(X_train_bow,Y_train)     #Train the model on training data

y_pred = rf.predict(X_test_bow)  # Predict the lables for the test data
accuracy_score(Y_test,y_pred)

0.8447671507260891

- What is a Decision Tree?
**A Decision Tree is a supervised learning algorithm used for classification and regression tasks. It works like a flowchart, making decisions based on data features step by step.**

Think of it as a series of "Yes" or "No" questions that lead to a final decision.

In [98]:
                #  (Root Node)
                #      "worst" in review?
                #     /        \
                # Yes (0)     No
                #          "amazing" in review?
                #         /        \
                #    Yes (1)     No
                #             "not" in review?
                #             /       \
                #         Yes (0)    No (1)


cv = CountVectorizer(max_features = 3000)
X_train_bow = cv.fit_transform(X_train['review']).toarray()
X_test_bow = cv.transform(X_test['review']).toarray()

rf = RandomForestClassifier()

rf.fit(X_train_bow,Y_train)
y_pred = rf.predict(X_test_bow)
accuracy_score(Y_test,y_pred)

0.8382573860791187

## N-grams technique

In [99]:
cv = CountVectorizer(ngram_range=(1,2),max_features=5000)
X_train_bow = cv.fit_transform(X_train['review']).toarray()
X_test_bow = cv.transform(X_test['review']).toarray()

rf = RandomForestClassifier()

rf.fit(X_train_bow,Y_train)
y_pred = rf.predict(X_test_bow)
accuracy_score(Y_test,y_pred)

0.8412618928392589

## TFIdf Technique

In [100]:
from sklearn.feature_extraction.text import TfidfVectorizer
tf = TfidfVectorizer()

In [101]:
X_train = tf.fit_transform(X_train['review']).toarray()
X_test = tf.transform(X_test['review'])


In [103]:
rf = RandomForestClassifier()
rf.fit(X_train,Y_train)
y_pred = rf.predict(X_test)
accuracy_score(Y_test,y_pred)

0.8457686529794692

## Word2vec Technique

In [104]:
import gensim
from gensim.models import Word2Vec,KeyedVectors
from nltk import sent_tokenize
from gensim.utils import simple_preprocess

In [106]:
story= []
for doc in df['review']:
  for sent in sent_tokenize(doc):
    story.append(simple_preprocess(sent))

In [107]:
model = gensim.models.Word2Vec(
    window=10,
    min_count=2,

)

In [108]:
model.build_vocab(story)

In [109]:
model.train(story,total_examples=model.corpus_count,epochs=model.epochs)

(5849776, 6186875)

In [110]:
len(model.wv.index_to_key)

31845

- I have to convert all the reviews in the form of vector , now I have converted word to vectors

##Average word2vec

- Calculate the average Word2Vec vector for a document (review).

In [111]:
def document_vector(doc):
  #remove out-of-vocabulary words
  doc = [word for word in doc.split() if word in model.wv.index_to_key]
  return np.mean(model.wv[doc],axis=0)

In [112]:
document_vector(df['review'].values[0])

array([-0.08900905,  0.44647878,  0.18740438, -0.05021271,  0.01453332,
       -0.7024713 , -0.00817924,  0.70281637, -0.33281264, -0.03183129,
       -0.14872539, -0.47340715,  0.05105571,  0.12640195,  0.22089309,
       -0.11844123,  0.09139479, -0.44949394, -0.15069416, -0.6272576 ,
        0.2187014 , -0.01559967,  0.13890228, -0.27303746, -0.19398989,
       -0.04945421, -0.2963386 , -0.3253056 , -0.27124274,  0.02985763,
        0.44997695, -0.01134141,  0.06338656, -0.25303057, -0.4053025 ,
        0.2899961 ,  0.04780946, -0.31987947, -0.14437036, -0.8589746 ,
       -0.01893315, -0.3354594 , -0.1939438 ,  0.08751079,  0.2891081 ,
       -0.14886448, -0.34866762, -0.1063443 ,  0.24868883,  0.31416097,
        0.17210905, -0.3877612 , -0.31171045, -0.22567876, -0.2648407 ,
        0.1126697 ,  0.37305033,  0.14940341, -0.20613377,  0.1486751 ,
       -0.14443593,  0.19895756, -0.1491931 , -0.05554924, -0.3066067 ,
        0.3943543 ,  0.21360993,  0.3246249 , -0.6896373 ,  0.37

In [113]:
from tqdm import tqdm

In [114]:
X =[]
for doc in tqdm(df['review'].values):
  X.append(document_vector(doc))

100%|██████████| 9983/9983 [05:21<00:00, 31.09it/s]


In [117]:
X = np.array(X)
X.shape

(9983, 100)

In [118]:
from sklearn.preprocessing import LabelEncoder
encoder = LabelEncoder()
Y = encoder.fit_transform(df['sentiment'])

In [119]:
Y

array([1, 1, 1, ..., 0, 0, 1])

In [120]:
from sklearn.model_selection import train_test_split
X_train,X_test,Y_train,Y_test = train_test_split(X,Y,test_size=0.2,random_state=1)

In [121]:
from sklearn.ensemble import RandomForestClassifier
rf = RandomForestClassifier()
rf.fit(X_train,Y_train)

y_pred = rf.predict(X_test)
accuracy_score(Y_test,y_pred)

0.771156735102654